src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Configuration/BaseProjectConfigurationDimensionProvider.cs

```diff
@@ -153,8 +153,58 @@ public IEnumerable<string> GetBestGuessDimensionNames(ImmutableArray<ProjectProp
+        public Task OnDimensionValueChangedAsync(ProjectConfigurationDimensionValueChangedEventArgs args)
+        {
+            if (StringComparers.ConfigurationDimensionNames.Equals(args.DimensionName, DimensionName))
+            {
+                if (args.Stage == ChangeEventStage.Before)
+                {
+                    switch (args.Change)
+                    {
+                        case ConfigurationDimensionChange.Add:
+                            return UpdateUnderLockAsync(args.Project, (msbuildProject, evaluatedPropertyValue) =>
+                                BuildUtilities.AppendPropertyValue(msbuildProject, evaluatedPropertyValue, PropertyName, args.DimensionValue));
+
+                        case ConfigurationDimensionChange.Delete:
+                            return UpdateUnderLockAsync(args.Project, (msbuildProject, evaluatedPropertyValue) =>
+                                BuildUtilities.RemovePropertyValue(msbuildProject, evaluatedPropertyValue, PropertyName, args.DimensionValue));
+
+                        case ConfigurationDimensionChange.Rename:
+                            // Need to wait until the core rename changes happen before renaming the property.
+                            break;
+                    }
+                }
+                else if (args.Stage == ChangeEventStage.After)
+                {
+                    // Only change that needs to be handled here is renaming configurations which needs to happen after all
+                    // of the core changes to rename existing conditions have executed.
+                    if (args.Change == ConfigurationDimensionChange.Rename)
+                    {
+                        return UpdateUnderLockAsync(args.Project, (msbuildProject, evaluatedPropertyValue) =>
+                            BuildUtilities.RenamePropertyValue(msbuildProject, evaluatedPropertyValue, PropertyName, args.OldDimensionValue, args.DimensionValue));
+                    }
+                }
+            }
+
+            return Task.CompletedTask;
+            
+            async Task UpdateUnderLockAsync(UnconfiguredProject project, Action<ProjectRootElement, string> action)
+            {
+                ConfiguredProject? configuredProject = await project.GetSuggestedConfiguredProjectAsync();
+
+                Assumes.NotNull(configuredProject);
+
+                await ProjectAccessor.OpenProjectForUpgradeableReadAsync(configuredProject, evaluatedProject =>
+                {
+                    string evaluatedPropertyValue = evaluatedProject.GetPropertyValue(PropertyName);
+
+                    return ProjectAccessor.OpenProjectXmlForWriteAsync(project, msbuildProject =>
+                    {
+                        action(msbuildProject, evaluatedPropertyValue);
+                    });
+                });
+            }
+        }    
 
         /// <summary>
         /// Gets the value for the specified property of the specified project.
```

Comment:
Can any of the protected methods below here be removed now?

Reply:
I just double checked and as it stands they're all used, but your question prompted me to wonder if we cannot flatten the types further, and I think we can. The override in `TargetFrameworkProjectConfigurationDimensionProvider` doesn't appear to be doing anything different from the base. Investigating.